### ISSUE TITLE
Seeking advice for fitting Generalised Lotka-Volterra Model.

### ISSUE THREAD

**Issue by mathematicsofpaul** [dynamicslab/pysindy#157]
Dear dynamicslab, 

I was seeking advice on how to fit the following form for each trajectory of x_i (where the index i represents the i th trajectory). The issue that I am facing is that I wish to enforce the constraint that each respective ODE is of the form x_i*( linear combination of trajectories). After reading further about the libraries such as `ps.PolynomialLibrary`, I still do not arrive at what I want. This is because, by using the library, I end up with terms that are not multiples of the respective x_i. I end up with other terms such as  x_j*x_{j'} along with the occasional x_i times other terms. Any assistance would be greatly appreciated. 

Kind Regards, Paul. 

![image](https://user-images.githubusercontent.com/31915369/151361609-1c5ac717-097e-4cdc-8510-b5f52676f07e.png)

P.S one could use a log trick to get a solution for a linear combination and then undo the log. Unfortuately, some of my trajectories are negative.

**Comment by mathematicsofpaul** [comment]
Not quite. I was looking to fit for each ode {x_i}` the equation {x_i}`=r_i*x_i+a_1*x_i*x_1+a_2*x_i*x_2+a_3*x_i*x_3+...+a_n*x_i*x_n. Or rather, {x_i}`=x_i times (r_i + a_i1*x_1+a_i2*x_2+...+a_in*x_n).

**Comment by akaptano** [comment]
Somehow I managed to misread your gigantic image. Probably the best way to do this is to use the ConstrainedSR3 library with constraints that the linear coefficient matrix is diagonal and another constraint for the quadratic part of the coefficients. Let me know if you need some more help with this.

**Comment by mathematicsofpaul** [comment]
Thank you for getting back to me so quickly! I went ahead and found your [tutorial](https://youtu.be/cPI466f2_4c?t=383) related to the ConstrainedSR3 library. I have just got a few questions related to this video.

How does setting the following `constraint_lhs` encode the opposite signs? What does 1 mean ? 
```
constraint_lhs[0,1] = 1
constraint_lhs[0,2] = 1
```

Secondly, how does the indexing work when setting the x coefficient to 28 for the second equation? 

`constraint_lhs[1,1 + n_features]=1`

Picture for reference.
![image](https://user-images.githubusercontent.com/31915369/151457273-50b32d89-6670-42d4-b943-6e990b44d888.png)

**Comment by akaptano** [comment]
You will have to play around with to figure out what is going on. The comments in the code above tries to explain that constraint_lhs[0, 1] = 1 and constraint_lhs[0, 2] = 1 (all other terms in the 0th row are zero) means that for the first variable equation (i.e. the equation for x' above) you want 
0 * 1 + 1 * x + 1 * y + 0 * z + 0 * x^2 + ... + 0 * z^2 = 0 = constraint_rhs[0]

In other words, you want x = -y in the equation for x'. 

Similarly, constraint_lhs[1, 1+n_features] = 1 means that for the y' equation you want
0 * 1 + 1 * x + 0 * y + 0 * z + .... + 0 * z^2 = constraint_rhs[1] = 28

In other words, you want x = 28 in the y' equation.

**Comment by mathematicsofpaul** [comment]
Thank you, this was very illuminating. 

> ... use the ConstrainedSR3 library with constraints that the linear coefficient matrix is diagonal and another constraint for the quadratic part of the coefficients. Let me know if you need some more help with this.

Actually some assistance would be greatly appreciated haha. 

Could you say more about how this could be done for my case? In particular, I am not sure where I should include a diagonal matrix for the linear term (and the other matrix for the quadratic terms). It appears that I should modify the theta matrix but how does that involve the 'constraint` matrices I mentioned above? 

Initially, I thought it would be pretty easy with some clever construction of the constraint equation to zero out non x_i linear coefficients for each respective {x_i}'. However, it seems non-trival. The ultimate goal would be to set 0=x_1=x_2=...=x_{i-1}=x_{i+1}=x_{n}. However, this is not clear with the contraint matrix/equation.

**Comment by mathematicsofpaul** [comment]
Thank you for making the time to write this Alan.

By setting the coefficient to 1, wouldn't that mean that the coefficient of say x in x' is unlearnable and restricted to a value of 1 only? In my case, I was particular interested in learning that coefficient whilst zero-ing out others.

**Comment by akaptano** [comment]
No, because I was sloppy earlier. What the constraints are really doing is:

0 * coefficient_of_1 + 1 * coefficient_of_x + 1 * coefficient_of_y + ... + 0 * coefficient_of_z^2 = 0 = constraint_rhs[0]

In other words, you want the **coefficient of  x = - the coefficient y** in the equation for x'.

and so forth

**Comment by akaptano** [comment]
Oh crap, no you are right, 
constraint_lhs[0, 1] = 1, constraint_rhs[0] = 1 (set coefficient of x in x' equation to 1)
sets the coefficient to identity, so it is unlearnable. This is fixable by simple omitting this line (and the same ones for the other equations). That way, those coefficients are unconstrained. So you have 6 constraints, corresponding to the off-diagonals of the linear system matrix. Let me know if that doesn't make sense.

**Comment by mathematicsofpaul** [comment]
This makes sense Alan. I had other questions related to this topic. 

1. Is there a way of printing out the library terms used without performing a model fit first? 

2. Since my group is using pysindy for the purpose of large volume trajectory modeling, did you have any insights on how many trajectories can be modelled at one time (we were thinking of roughly 200 trajectories to be modelled)? At the moment, I have only seen pysindy perform on small systems such as lorenz systems where only n=3 "coordinates" are modelled. Is there a limit?

**Comment by akaptano** [comment]
1. Yes, if you have a SINDy library defined you can also do library.fit(x) and then library.get_feature_names().

2. Depends on what you mean here by "trajectories" since the Lorenz system is 3D system, but you can easily use 1000 different Lorenz trajectories (starting from different initial conditions, with different time bases, etc.). So the answer is yes you can fit SINDy with many different trajectories of a system. For this, check out the multiple_trajectories functionality that is shown in the Example 1 Jupyter notebook. 

If you meant that you are trying to a model a 200-dimensional system, this is also "in principle" doable with SINDy but it is basically impossible. You are trying to regress onto a 200-dimensional coupled system of ODEs, which could easily have 10,000 free parameters. If you clarify what you need I can provide some more recommendations.

**Comment by mathematicsofpaul** [comment]
Thank you Alan! 

To clarify, indeed I am looking to model a 200 dimensional system with polynomials of order 2 (interaction included) with the restriction limited to the equations I posted earlier a.k.a with the constraints to encode the GLV equational form. Hence, 200 free parameters per dimension. This should roughly result in 201*201 ~ 40000 free parameters :(.

**Comment by mathematicsofpaul** [comment]
This is really fascinating. Who would've have known fitting a GLV model could be so involved haha. When we encode these stability conditions, how do we check to ensure that they are inline with the underlying structure of the ODEs? In other words, could enforcing these stability conditions artificially prevent us from fitting to the true form of the underlying ODEs?

**Comment by akaptano** [comment]
To clarify, what I am saying is that you need the stability constraints that **are** in line with the underlying structures of the ODEs. In other words, a generic 200D quadratic nonlinearity model will not be stable for any initial condition (although you may not care about that... you may only want stability within some region of the origin in which case linear stability is enough). I'm saying that **if your 200D GLV model is stable for any initial condition, this actually is an extremely strong constraint on the structure of the coefficients in the model** and you need to build this constraint into the SINDy process to have any hope of reproducing this (important) feature.

**Comment by mathematicsofpaul** [comment]
Hi Alan, I have been making good progress on setting the constraints you helped me with earlier in the thread. Did you have any insights on how to include an intercept term in a custom library? I am struggling to find an example of how to do this. 

```
library_functions = [lambda x:x, lambda x, y:x * y, lambda x:x ** 2]
library_function_names = [lambda x:x, lambda x, y:x + y, lambda x:x + x]
sindy_library = ps.CustomLibrary(library_functions=library_functions,
                                 function_names=library_function_names)
```
will give me something like: 

```
sindy_library.fit(X)
sindy_library.get_feature_names() 

['x0', 'x1', 'x0x1', 'x0x0', 'x1x1']
```
Looking for something like `['1', 'x0', 'x1', 'x0x1', 'x0x0', 'x1x1']`.

**Comment by akaptano** [comment]
That one is easy. Use include_bias=True in the CustomLibrary :)

**Comment by akaptano** [comment]
Note that this will change the indexing of all your constraints by one

**Comment by akaptano** [comment]
By the way, I looked up the GLV wikipedia page and this system is stable if the alpha_{ij} is a negative definite matrix or the elements of alpha_{ij} <= 0 are all zero. This has weaker stability requirements than usual because this equation is modelling populations... i.e. x_i(t) >= 0 for all time is assumed. In this case:
dE/dt = x_i * dx_i/dt = r_i * x_i^2 + alpha_{ij} x_i^2 x_j,
and, since x_j >= 0 for all time, if alpha_{ij} is negative definite (and non-vanishing), this cubic term is always negative everywhere in the state space. We conclude the system is globally stable.

**Comment by mathematicsofpaul** [comment]
![image](https://user-images.githubusercontent.com/31915369/152458154-f5bc0fd5-4612-4ced-bb5b-a4007b51ce18.png)

Hey Alan, 

I am getting some interesting behaviour from the constrainedSR3 function. In particular, I am getting an output that is not expected given my constraints. In particular, for my first result in model.print(), the term OTU84 is not supposed to appear since only terms of factor OTU25 are meant to show up. Similarly, for the second result, I am getting an intercept term. If one looks at my constraint matrix, it should not show up. Rather strange ... why are these terms showing up. I am missing something?  

P.S The fit is for a simple 2D system where the constraint matrix coefficients should be ordered as 
`['1', 'OTU25', 'OTU84', 'OTU25^2', 'OTU25 OTU84', 'OTU84^2']` for the first dimension. Then, of course, again for the second dimension as,`['1', 'OTU25', 'OTU84', 'OTU25^2', 'OTU25 OTU84', 'OTU84^2']`.

**Comment by akaptano** [comment]
Try constraint_feature = "target"

**Comment by mathematicsofpaul** [comment]
Hi Alan! Would you have any advice on a direction to resolving this? Seems like a vanishing value problem but im not sure where to start..

```
sim = model.simulate(initial_values, t = t_test, integrator="solve_ivp")

lsoda--  warning..internal t (=r1) and h (=r2) are
       such that in the machine, t + h = t on the next step  
       (h = step size). solver will continue anyway
      in above,  r1 =  0.6559806598148D+01   r2 =  0.4095581883771D-15
. . . .
ValueError: Input contains NaN, infinity or a value too large for dtype('float64').
```

**Comment by akaptano** [comment]
Anyways if you post a bit more of your progress/setup I might be able to comment further.

**Comment by mathematicsofpaul** [comment]
Thank you Alan, in my case it is only 5 variables that I am interesting in modeling.

```
sindy_library = ps.PolynomialLibrary(degree=2)
sindy_opt = ps.ConstrainedSR3(
    threshold=threshold,
    nu=1,
    max_iter=10000,
    constraint_lhs = constraint_matrix,
    constraint_rhs = constraint_zeros,
    constraint_order = "target",
    tol = 1e-3,
    initial_guess=initial_guess
)  

model = ps.SINDy(
    optimizer=sindy_opt,
    feature_library=sindy_library,
    differentiation_method=ps.FiniteDifference(drop_endpoints=True),
    feature_names=otus
)

model.fit(X, t=t)
model.print()

(OTU25)' = 0.812 OTU25 + -0.110 OTU25^2 + 0.102 OTU25 OTU98 + -1.729 OTU25 OTU130
(OTU84)' = -4.500 OTU84 + -0.706 OTU25 OTU84 + -0.319 OTU84^2 + 0.950 OTU84 OTU98 + 1.586 OTU84 OTU130 + -0.414 OTU84 OTU140
(OTU98)' = 5.574 OTU98 + -1.095 OTU98^2 + -1.745 OTU98 OTU130 + 2.738 OTU98 OTU140
(OTU130)' = -5.341 OTU130 + 1.188 OTU98 OTU130 + 1.170 OTU130^2 + -2.792 OTU130 OTU140
(OTU140)' = 2.420 OTU140 + -1.887 OTU98 OTU140 + 4.297 OTU130 OTU140 + 2.584 OTU140^2
```

With the family of functions: 

```
model.get_feature_names() 
['1',
 'OTU25',
 'OTU84',
 'OTU98',
 'OTU130',
 'OTU140',
 'OTU25^2',
 'OTU25 OTU84',
 'OTU25 OTU98',
 'OTU25 OTU130',
 'OTU25 OTU140',
 'OTU84^2',
 'OTU84 OTU98',
 'OTU84 OTU130',
 'OTU84 OTU140',
 'OTU98^2',
 'OTU98 OTU130',
 'OTU98 OTU140',
 'OTU130^2',
 'OTU130 OTU140',
 'OTU140^2']
```

**Comment by mathematicsofpaul** [comment]
Sorry exactly what did you mean by negative definite? I was excited by the constraints you reccomended earlier, but unfortunately my dimensions vary both negatively and positively, so im not sure how the constrainted could be adjusted. In terms of my constraints, they are only constraints that force the intercepts terms to be zero and force the resulting equations to be of GLV form.

**Comment by mathematicsofpaul** [comment]
I have been meditating on what we have been discussing for the past few days, and I have come back with a few queries. 

1. I read online that the importance of having stable models is that no matter what initial conditions we begin with, we will with some degree of certainty be able to determine the behaviour of the system later in time. Is this the type of stability that we are discussing, and why we are stressing the building of stable models so much? Or is there something different in the numerical setting that stresses building numerical models? 
2. The constraint alpha_{ij} <=0 would get the job done for -inf < x_i(t) < inf. However, since we are working with GLVs, the community/interaction matrix A needs to be of mixed positive and negative value in order for us to interpret the "interactions" between the dimensions. At this point it is looking like I would have to derive constraints that allow for negative and positive a_ij for -inf < x_i(t) < inf. Did you have any ideas (or where to read) as to where I could look to derive these constraints myself? 
3. Despite the fact that my system is unstable as shown by the discussions above, does this mean that my results from model.fit() a.k.a. 
```
model.fit(X, t=t)
model.print()

(OTU25)' = 0.812 OTU25 + -0.110 OTU25^2 + 0.102 OTU25 OTU98 + -1.729 OTU25 OTU130
(OTU84)' = -4.500 OTU84 + -0.706 OTU25 OTU84 + -0.319 OTU84^2 + 0.950 OTU84 OTU98 + 1.586 OTU84 OTU130 + -0.414 OTU84 OTU140
(OTU98)' = 5.574 OTU98 + -1.095 OTU98^2 + -1.745 OTU98 OTU130 + 2.738 OTU98 OTU140
(OTU130)' = -5.341 OTU130 + 1.188 OTU98 OTU130 + 1.170 OTU130^2 + -2.792 OTU130 OTU140
(OTU140)' = 2.420 OTU140 + -1.887 OTU98 OTU140 + 4.297 OTU130 OTU140 + 2.584 OTU140^2
```
are meaningless since the system is unstable? In other words, would the interaction value of 0.102 between OTU25 & OTU98 carry no meaning since the system is unstable?

**Comment by akaptano** [comment]
Responding to each of these:
1. Global stability here means that if your system is unstable, there are some initial trajectories such that it will blow up to infinity. You can always determine the behavior of the governing equations later in time by integrating the equations with solve_ivp. The problem is that sometimes that behavior is flying to infinity because your identified model coefficients aren't obeying the symmetries of the system. You are right that you may be able to get away with this sometimes, although it seems you already finding issues with your 5D model.

2. You can have positive and negative values in alpha_{ij}, as long as it is a negative definite matrix. This is required even in the restricted domain of interest where 0 <= x_i(t) < inf. This follows immediately by computing the energy and its derivative as I did above... there is a cubic term in the derivative that goes like  alpha_{ij}x_i^2x_j. If x_j > 0 and alpha_{ij} is negative definite, this cubic term is always stabilizing. I do not have more references (for a more advanced reference see "Evolutionary games and population dynamics" by Hofbauer) beyond the ones above because I am not a GLV model expert. 

3. You are correct, your results are not meaningless. Finding the interaction coefficients is quite useful. You can plot your fit against the true x_dot data, and it is likely the SINDy model fits the derivative data quite well. The coefficients are likely very close to correct. This is valuable in its own right -- you have identified a set of equations that is approximately correct for the system. The issue is just that your model weakly breaks a fundamental symmetry of the system, and therefore if you integrate the x_dot SINDy equations, you can get unstable trajectories. My point was not that you **need** to fix this issue, but merely that you **need** to fix this issue if you want to integrate the equations in model.simulate, which you specifically asked about earlier. 

I think you are off to a good start here, so I am going to close this issue :)

Best,
Alan
